pkg/ext/stores/tokens/tokens.go

```diff
@@ -0,0 +1,1321 @@
+	token.Status.LastUpdateTime = t.timer.Now()
+
+	rest.FillObjectMetaSystemFields(token)
+
+	secret, err := secretFromToken(token, nil, nil)
+	if err != nil {
+		return nil, apierrors.NewInternalError(fmt.Errorf("failed to convert token %s for storage: %w",
+			token.Name, err))
+	}
+
+	// Check the user's accessibility of the cluster the token is limited to, if such is done.
+	if token.Spec.ClusterName != "" {
+		if err := t.IsClusterAccessible(ctx, token.Spec.ClusterName); err != nil {
+			return nil, err
+		}
+	}
+
+	// Save new secret
+	newSecret, err := t.secretClient.Create(secret)
+	if err != nil {
+		if apierrors.IsAlreadyExists(err) {
+			// can happen despite the early check for a pre-existing secret.
+			// something else may have raced us while the secret was assembled.
+			return nil, err
+		}
+		return nil, apierrors.NewInternalError(fmt.Errorf("failed to store token %s: %w",
+			token.Name, err))
+	}
+
+	// Read changes back to return what was truly created, not what we thought we created
+	newToken, err := tokenFromSecret(newSecret)
+	if err != nil {
+		// An error here means that something broken was stored. Do not
+		// leave that broken thing behind.
+		t.secretClient.Delete(TokenNamespace, secret.Name, &metav1.DeleteOptions{})
+
+		// And report what was broken
+		return nil, apierrors.NewInternalError(fmt.Errorf("failed to regenerate token %s: %w",
+			token.Name, err))
+	}
+
+	// users don't care about the hashed value
+	newToken.Status.TokenHash = ""
+	newToken.Status.TokenValue = tokenValue
+	return newToken, nil
+}
+
+func (t *Store) delete(ctx context.Context, token *ext.Token, options *metav1.DeleteOptions) error {
+	user, err := t.auth.UserName(ctx)
+	if err != nil {
+		return apierrors.NewInternalError(err)
+	}
+	if !userMatch(user, token) {
+		return apierrors.NewNotFound(GVR.GroupResource(), token.Name)
+	}
+
+	return t.SystemStore.Delete(token.Name, options)
+}
+
+func (t *SystemStore) Delete(name string, options *metav1.DeleteOptions) error {
+	err := t.secretClient.Delete(TokenNamespace, name, options)
+	if err == nil {
+		return nil
+	}
+	if apierrors.IsNotFound(err) {
+		return nil
+	}
+	return apierrors.NewInternalError(fmt.Errorf("failed to delete token %s: %w", name, err))
+}
+
+func (t *Store) get(ctx context.Context, name string, options *metav1.GetOptions) (*ext.Token, error) {
+	// note: have to get token first before we can check for a user mismatch
+	token, err := t.SystemStore.Get(name, options)
+	if err != nil {
+		return nil, err
+	}
+
+	user, err := t.auth.UserName(ctx)
+	if err != nil {
+		return nil, apierrors.NewInternalError(err)
+	}
+	if !userMatch(user, token) {
+		return nil, apierrors.NewNotFound(GVR.GroupResource(), name)
+	}
+
+	return token, nil
+}
+
+func (t *SystemStore) Get(name string, options *metav1.GetOptions) (*ext.Token, error) {
+	// Core token retrieval from backing secrets
+	currentSecret, err := t.secretClient.Get(TokenNamespace, name, *options)
```

Comment:
should we use a cache here? In general, should we use caches for getting/listing in the stores? We would increment the number of requests to the apiserver if we don't use caches

Reply:
To be done.

Reply:
As of commit 46a375518f8bdb4d89a9dfbb118e1190e82b498a (end of rebased commit sequence) the store uses caches for access to users and backing secrets, where possible (The store performs actions like Update, Delete, Patch which are not supported by the caches).